Sources/TensorFlow/Core/LazyTensorOperation.swift

```diff
@@ -725,3 +724,70 @@ extension LazyTensorOperation: CustomStringConvertible {
         return desc
     }
 }
+
+extension LazyTensorOperation {
+    /// Returns the materialized value at the given output `index`.
+    func materialized(index: Int) -> TFETensorHandle {
+        precondition(index < outputCount)
+        return materialized()[index]
+    }
+
+    /// Materializes all the outputs.
+    func materialized() -> [TFETensorHandle] {
+        // Return materialized outputs if any.
+        if let outputs = outputs { return outputs }
+
+        LazyTensorOperation.materializeLiveTensors(self)
+
+        // Our outputs should have been updated by now. Otherwise,
+        // something terrible happened!
+        precondition(outputs != nil, "Materialization failed!")
+        return outputs!
+    }
+
+    /// Converts symbolic tensor inputs to concrete inputs if the
+    /// associated `LazyTensorOperation` has been materialized.
+    private func maybeMaterializeInputs() {
+        func maybeMaterialized(lazyTensor: LazyTensor) -> LazyTensor {
+            let handle = lazyTensor.handle
+            if case let LazyTensor.Handle.symbolic(lazyOp, index, _) = handle {
+                if let outputs = lazyOp.outputs {
+                    return LazyTensor(_materialized: outputs[index])
+                }
+            }
+            return lazyTensor
+        }
+
+        func maybeMaterialized(input: Input) -> Input {
+            switch input {
+            case .single(let h):
+                return Input.single(maybeMaterialized(lazyTensor: h))
+            case .list(let elements):
+                return Input.list(elements.map { maybeMaterialized(lazyTensor: $0) })
+            }
+        }
+        inputs = inputs.map { maybeMaterialized(input: $0) }
+    }
+
+    private static func materializeLiveTensors(_ lazyOp: LazyTensorOperation) {
```

Comment:
Since this takes a `Self`, why not define it as an instance method?

Reply:
Done.